### PR TITLE
KAFKA-8436: use automated protocol for AddOffsetsToTxn 

### PR DESCRIPTION
**Pull request by abbccdda**
In an effort to replace the protocol of AddOffsetsToTxn API to automated.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
Retest this please

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -50,7 +50,7 @@ import org.apache.kafka.common.internals.FatalExitError
...
+import org.apache.kafka.common.message.{AddOffsetsToTxnResponseData, AlterPartitionReassignmentsResponseData, CreateAclsResponseData, CreatePartitionsResponseData, CreateTopicsResponseData, DeleteAclsResponseData, DeleteGroupsResponseData, DeleteRecordsResponseData, DeleteTopicsResponseData, DescribeAclsResponseData, DescribeGroupsResponseData, DescribeLogDirsResponseData, EndTxnResponseData, ExpireDelegationTokenResponseData, FindCoordinatorResponseData, HeartbeatResponseData, InitProducerIdResponseData, JoinGroupResponseData, LeaveGroupResponseData, ListGroupsResponseData, ListPartitionReassignmentsResponseData, OffsetCommitRequestData, OffsetCommitResponseData, OffsetDeleteResponseData, RenewDelegationTokenResponseData, SaslAuthenticateResponseData, SaslHandshakeResponseData, StopReplicaResponseData, SyncGroupResponseData, UpdateMetadataResponseData}
```
Comment (abbccdda) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Adding `AddOffsetsToTxnResponseData` in the front

**Comment by abbccdda** [comment]
@mimaison Could you take a look for this one?

**Comment by abbccdda** [comment]
@mimaison Could you have a look for this one?

**Comment by abbccdda** [comment]
@mimaison Could you take a look for this one?

**Comment by abbccdda** [comment]
@mimaison Could you take a look for this one?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 2]
```diff
@@ -3258,12 +3259,16 @@ private void prepareAddOffsetsToTxnResponse(final Errors error,
...
-            assertEquals(producerEpoch, addOffsetsToTxnRequest.producerEpoch());
+            assertEquals(consumerGroupId, addOffsetsToTxnRequest.data.groupId());
+            assertEquals(transactionalId, addOffsetsToTxnRequest.data.transactionalId());
+            assertEquals(producerId, addOffsetsToTxnRequest.data.producerId());
+            assertEquals(producerEpoch, addOffsetsToTxnRequest.data.producerEpoch());
             return true;
-        }, new AddOffsetsToTxnResponse(0, error));
+        }, new AddOffsetsToTxnResponse(
+            new AddOffsetsToTxnResponseData()
+                .setThrottleTimeMs(0)
```
Comment (mimaison) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: Throttle time defaults to `0`, so this is not required

File: clients/src/main/java/org/apache/kafka/common/requests/AddOffsetsToTxnResponse.java [hunk 3]
```diff
@@ -16,86 +16,64 @@
...
-        this.throttleTimeMs = throttleTimeMs;
-        this.error = error;
+    public AddOffsetsToTxnResponse(AddOffsetsToTxnResponseData data) {
+        this.data = data;
     }
 
     public AddOffsetsToTxnResponse(Struct struct) {
-        this.throttleTimeMs = struct.get(THROTTLE_TIME_MS);
-        this.error = Errors.forCode(struct.get(ERROR_CODE));
+        this.data = new AddOffsetsToTxnResponseData(struct, (short) (AddOffsetsToTxnResponseData.SCHEMAS.length - 1));
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/common/requests/AddOffsetsToTxnResponse.java]: As far as I can tell, this constructor is not used anywhere. Can we delete it?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 4]
```diff
@@ -1630,7 +1638,7 @@ public void handleResponse(AbstractResponse response) {
             } else if (error == Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED) {
                 fatalError(error.exception());
             } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
-                abortableError(GroupAuthorizationException.forGroupId(builder.consumerGroupId()));
+                abortableError(new GroupAuthorizationException(builder.data.groupId()));
```
Comment (mimaison) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: Why are we changing this? The constructor with a single argument takes a message and not the groupId

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks for the PR, I've left a few small comments

**Comment by abbccdda** [comment]
@mimaison Thanks for the review! Addressed

**Comment by mimaison** [comment]
ok to test

**Comment by abbccdda** [comment]
@mimaison Looks 2 checks passed

**Comment by abbccdda** [comment]
@mimaison Does the change look good to you?

**Review by mimaison — APPROVED** [review]

**Pull request merged**